File: ObjectiveTransition/ObjectiveTransition/CircularTransition.m
//
//  CircularTransition.m
//  ObjectiveTransition
//
//  Created by DEN on 07/08/17.
//  Copyright © 2017 14K. All rights reserved.
//

#import "CircularTransition.h"



@interface CircularTransition() <UIViewControllerAnimatedTransitioning>

@property (assign,nonatomic) double duration;
@property (strong,nonatomic) UIView *circle;

@end

@implementation CircularTransition


- (instancetype)init{
    self = [super init];
    
    self.duration = 0.5;
    self.circle = [[UIView alloc] init];
    self.transitionMode = present;
    self.circleColor = [UIColor whiteColor];
    self.startingPoint = CGPointZero;
    
    
    return self;
}


- (NSTimeInterval)transitionDuration:(id<UIViewControllerContextTransitioning>)transitionContext{
    return self.duration;
}

- (void)animateTransition:(id<UIViewControllerContextTransitioning>)transitionContext{
    
    UIView *containerView = transitionContext.containerView;
    
    if(self.transitionMode == present){
        
        UIView *presentedView;
        
        if((presentedView = [transitionContext viewForKey:UITransitionContextToViewKey])){
            
            CGPoint viewCenter = presentedView.center;
            CGSize viewSize = presentedView.frame.size;
            
            self.circle = [[UIView alloc] init];
            
            self.circle.frame = [self frameForCircleWithCenter:viewCenter size:viewSize startPoint:self.startingPoint];
            
            self.circle.layer.cornerRadius = self.circle.frame.size.height/2.0;
            
            self.circle.center = self.startingPoint;
            self.circle.backgroundColor = self.circleColor;
            
            self.circle.transform = CGAffineTransformMakeScale(0.001, 0.001);
            [containerView addSubview:self.circle];
            
            presentedView.center = self.startingPoint;
            presentedView.transform = CGAffineTransformMakeScale(0.001, 0.001);
            presentedView.alpha = 0.0;
            
            [containerView addSubview:presentedView];
            
            [UIView animateWithDuration:self.duration animations:^{
                
                self.circle.transform = CGAffineTransformIdentity;
                presentedView.transform = CGAffineTransformIdentity;
                presentedView.center = viewCenter;
                presentedView.alpha = 1.0;
                
            } completion:^(BOOL finished){
                [transitionContext completeTransition:finished];
            }];
            
        }
        
    }
    else{
        UITransitionContextViewKey transitionModeKey = (self.transitionMode == pop) ? UITransitionContextToViewKey : UITransitionContextFromViewKey;
        
        UIView *returningView;
        if((returningView = [transitionContext viewForKey:transitionModeKey])){
            CGPoint viewCenter = returningView.center;
            CGSize viewSize = returningView.frame.size;
            
            self.circle.frame = [self frameForCircleWithCenter:viewCenter size:viewSize startPoint:self.startingPoint];
             
            self.circle.layer.cornerRadius = self.circle.frame.size.height/2.0;
            
            self.circle.center = self.startingPoint;
            
            [UIView animateWithDuration:self.duration animations:^{
                
                self.circle.transform = CGAffineTransformMakeScale(0.001, 0.001);
                
                returningView.transform  = CGAffineTransformMakeScale(0.001, 0.001);
                returningView.center = self.startingPoint;
                returningView.alpha = 0.0;
                
                if(self.transitionMode == pop){
                    [containerView insertSubview:returningView belowSubview:returningView];
                    [containerView insertSubview:self.circle belowSubview:returningView];
                }
                
            
            } completion:^(BOOL finished){
                returningView.center = viewCenter;
                [returningView removeFromSuperview];
                [self.circle removeFromSuperview];
                [transitionContext completeTransition:finished];
            }];
            
        }
        
        
    }
    
}


- (CGRect)frameForCircleWithCenter:(CGPoint)center size:(CGSize)size startPoint:(CGPoint)startingPoint{
    
    
    CGFloat xLength = fmax(startingPoint.x, size.width - startingPoint.x);
    CGFloat yLength = fmax(startingPoint.y, size.height - startingPoint.y);
    
    CGFloat offsetVector = sqrt(xLength*xLength + yLength*yLength)*2;
    
    return CGRectMake(CGPointZero.x, CGPointZero.y, offsetVector, offsetVector);
    
}



- (double)duration{
    if((!_duration)||(_duration == 0)){
        _duration = 0.5;
    }
    
    return _duration;
}






@end

File: Circular Popup/CircularTransition.swift
//
//  CircularTransition.swift
//  Circular Popup
//
//  Created by DEN on 06/08/17.
//  Copyright © 2017 14K. All rights reserved.
//

import UIKit

class CircularTransition: NSObject {

    var circle = UIView()
    
    var startingPoint = CGPoint.zero{
        didSet{
            circle.center = startingPoint;
        }
    }
    var circleColor = UIColor.white
    var duration = 0.3
    
    enum CircularTransitionMode{
        case present
        case dismiss
        case pop
    }
    
    var transitionMode:CircularTransitionMode = .present
    
}


extension CircularTransition:UIViewControllerAnimatedTransitioning{
    func transitionDuration(using transitionContext: UIViewControllerContextTransitioning?) -> TimeInterval {
        return duration
    }
    func animateTransition(using transitionContext: UIViewControllerContextTransitioning) {
        let containerView = transitionContext.containerView
        
        if transitionMode == .present{
            if let presentedView = transitionContext.view(forKey: UITransitionContextViewKey.to){
                let viewCenter = presentedView.center
                let viewSize = presentedView.frame.size
                
                circle = UIView()
                
                circle.frame = frameForCircle(withViewCenter: viewCenter, size: viewSize, startPoint: startingPoint)
                circle.layer.cornerRadius = circle.frame.size.height / 2
                circle.center = startingPoint
                circle.backgroundColor = circleColor
                
                circle.transform = CGAffineTransform(scaleX: 0.001, y: 0.001)
                containerView.addSubview(circle)
                
                presentedView.center = startingPoint
                presentedView.transform = CGAffineTransform(scaleX: 0.001, y: 0.001)
                presentedView.alpha = 0.0
                
                containerView.addSubview(presentedView)
                
                UIView.animate(withDuration: duration, animations: {
                    self.circle.transform = CGAffineTransform.identity
                    presentedView.transform = CGAffineTransform.identity
                    presentedView.alpha = 1.0
                    presentedView.center = viewCenter
                }, completion:{ (success:Bool) in
                    transitionContext.completeTransition(success)
                })
            }
        }
        else{
            let transitionKey = (transitionMode == .pop) ? 	UITransitionContextViewKey.to : UITransitionContextViewKey.from
            
            if let returningView = transitionContext.view(forKey: transitionKey){
                let viewCenter = returningView.center
                let viewSize = returningView.frame.size
                
                circle.frame = frameForCircle(withViewCenter: viewCenter, size: viewSize, startPoint: startingPoint)
                
                circle.layer.cornerRadius = circle.frame.size.height/2
                circle.center = startingPoint
                
                UIView.animate(withDuration: duration, animations: {
                    self.circle.transform = CGAffineTransform(scaleX: 0.001, y: 0.001)
                    returningView.transform = CGAffineTransform(scaleX: 0.001, y: 0.001)
                    returningView.center = self.startingPoint
                    returningView.alpha = 0
                    
                    if self.transitionMode == .pop{
                        containerView.insertSubview(returningView, belowSubview: returningView)
                        containerView.insertSubview(self.circle, belowSubview: returningView)
                    }
                    
                }, completion:{ (success : Bool) in
                
                     returningView.center = viewCenter
                    returningView.removeFromSuperview()
                    self.circle.removeFromSuperview()
                    transitionContext.completeTransition(success)
                
            })
        }
    }
}
    
    func frameForCircle(withViewCenter viewCenter:CGPoint, size viewSize:CGSize, startPoint:CGPoint) -> CGRect{
        let xlength = fmax(startPoint.x,viewSize.width - startPoint.x)
        let yLength = fmax(startPoint.y, viewSize.height - startPoint.y)
        
        let offSetVector = sqrt(xlength*xlength + yLength*yLength)*2
        let size = CGSize(width: offSetVector, height: offSetVector)
        
        return CGRect(origin: CGPoint.zero, size:size)
    }
    
}













